File: challenge_1/src/main/java/com/ing/challenge/parser/CsvProductParser.java
package com.ing.challenge.parser;

import com.ing.challenge.model.Product;
import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVPrinter;
import org.apache.commons.csv.CSVRecord;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.Reader;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class CsvProductParser implements Parser<Product> {
    /**
     * {@inheritDoc}
     *
     * @param fileName name of the file
     * @return A {@link List} of Object of type T
     * @see Parser#read(File)
     */
    @Override
    public List<Product> read(String fileName) {
        try (Reader in = new FileReader(fileName)) {
            Iterable<CSVRecord> records = CSVFormat.DEFAULT
                    .withFirstRecordAsHeader()
                    .withIgnoreSurroundingSpaces()
                    .parse(in);
            return StreamSupport.stream(records.spliterator(), false)
                    .map(Product::fromCsv)
                    .collect(Collectors.toList());
        } catch (IOException e) {
            e.printStackTrace();
            return Collections.emptyList();
        }
    }

    /**
     * {@inheritDoc}
     *
     * @param file A {@link File} object relating to A file on the system
     * @return A {@link List} of Object of type T
     * @see Parser#read(String)
     */
    @Override
    public List<Product> read(File file) {
        return read(file.getAbsolutePath());
    }

    @Override
    public boolean write(File file, Product product) {
        return write(file, List.of(product));
    }

    @Override
    public boolean write(File file, List<Product> products) {
        boolean hasFailed = false;
        try (CSVPrinter csvPrinter = CSVFormat.DEFAULT
                .withHeader("productId", "name", "description", "price", "category")
                .print(file, StandardCharsets.UTF_8)) {
            String[][] data = new String[products.size()][];
            int x = 0;
            for (var p : products) {
                data[x++] = new String[]{
                        String.valueOf(p.productId()),
                        p.name(),
                        p.description(),
                        String.valueOf(p.price()),
                        p.category()
                };
            }
            csvPrinter.printRecords((Object[]) data);
            csvPrinter.flush();
        } catch (IOException e) {
            e.printStackTrace();
            hasFailed = true;
            //throw e;
        }
        return !hasFailed;
    }
}

File: vincent/challenge_1K/src/main/kotlin/com/ing/challenge/parser/CsvProductParser.kt
package com.ing.challenge.parser

import com.ing.challenge.model.Product
import org.apache.commons.csv.CSVFormat
import org.apache.commons.csv.CSVRecord
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.nio.charset.StandardCharsets

class CsvProductParser : Parser<Product> {

    override fun read(fileName: String): List<Product> {
        return try {
            FileReader(fileName).use { reader ->
                val records: Iterable<CSVRecord> = CSVFormat.DEFAULT
                        .withFirstRecordAsHeader()
                        .withIgnoreSurroundingSpaces()
                        .parse(reader)
                records.map { record: CSVRecord -> Product.fromCsv(record) }
            }
        } catch (e: IOException) {
            e.printStackTrace()
            return emptyList()
        }
    }


    override fun read(file: File): List<Product> {
        return read(file.absolutePath)
    }

    override fun write(file: File, obj: Product): Boolean {
        return write(file, listOf(obj))
    }

    override fun write(file: File, objects: List<Product>): Boolean {
        val result: Result<Unit> = CSVFormat.DEFAULT
                .withHeader("productId", "name", "description", "price", "category")
                .print(file, StandardCharsets.UTF_8).use { csvPrinter ->
                    val productList = objects.fold(mutableListOf<Array<String>>()) { acc, product ->
                        val array = with(product) {
                            arrayOf(
                                    productId.toString(),
                                    name,
                                    description,
                                    price.toString(),
                                    category,
                            )
                        }
                        acc.apply { add(array) }
                    }
                    kotlin.runCatching {
                        csvPrinter.printRecords(productList)
                        csvPrinter.flush()
                    }
                }
        return result.isSuccess
    }
}
